From 940cc27370ff6f55593953d88c9beeccb02a319d Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Wed, 7 Aug 2019 20:18:12 +0200
Subject: [PATCH 1/1] efi_loader: EFI_HII_DATABASE_PROTOCOL.SetKeyboardLayout()

Return correct status code when executing the SetKeyboardLayout() service
of the EFI_HII_DATABASE_PROTOCOL.

The actual changing of the keyboard layout is not yet implemented.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_hii.c | 30 +++++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/lib/efi_loader/efi_hii.c b/lib/efi_loader/efi_hii.c
index 77e330285a..f86bc9fcd2 100644
--- a/lib/efi_loader/efi_hii.c
+++ b/lib/efi_loader/efi_hii.c
@@ -752,13 +752,41 @@ found:
 	return EFI_EXIT(EFI_SUCCESS);
 }
 
+/**
+ * set_keyboard_layout() - set the current keyboard layout
+ *
+ * This function implements the SetKeyboardLayout() service of the
+ * EFI_HII_DATABASE_PROTOCOL.
+ *
+ * See the Unified Extensible Firmware Interface (UEFI) specification for
+ * details.
+ *
+ * @this:	HII database protocol
+ * @key_guid:	GUID of the keyboard layout to be set
+ */
 static efi_status_t EFIAPI
 set_keyboard_layout(const struct efi_hii_database_protocol *this,
 		    efi_guid_t *key_guid)
 {
+	struct efi_keyboard_layout_data *layout_data;
+	efi_status_t ret = EFI_NOT_FOUND;
+
 	EFI_ENTRY("%p, %pUl", this, key_guid);
 
-	return EFI_EXIT(EFI_NOT_FOUND);
+	if (!key_guid) {
+		ret = EFI_INVALID_PARAMETER;
+		goto out;
+	}
+
+	list_for_each_entry(layout_data, &efi_keyboard_layout_list, link_sys) {
+		if (!guidcmp(&layout_data->keyboard_layout.guid, key_guid)) {
+			/* TODO: actually set current keyboard layout */
+			ret = EFI_SUCCESS;
+			goto out;
+		}
+	}
+out:
+	return EFI_EXIT(ret);
 }
 
 static efi_status_t EFIAPI
-- 
2.20.1

